src/charm.py

```diff
@@ -0,0 +1,1624 @@
+            mount_suffix=event.mount_suffix,
+            egress_subnets=event.egress_subnets,
+            nonce=event.nonce,
+        )
+
+    def _on_vault_kv_client_detached(self, event: VaultKvClientDetachedEvent):
+        label = self._get_vault_kv_secret_label(unit_name=event.unit_name)
+        self._remove_juju_secret_by_label(label=label)
+
+    def _configure_pki_secrets_engine(self) -> None:
+        """Configure the PKI secrets engine."""
+        if not self.unit.is_leader():
+            logger.debug("Only leader unit can handle a vault-pki certificate request")
+            return
+        vault = self._get_active_vault_client()
+        if not vault:
+            logger.debug("Failed to get initialized Vault")
+            return
+        if not self._tls_certificates_pki_relation_created():
+            logger.debug("TLS Certificates PKI relation not created, skipping")
+            return
+        if not self._common_name_config_is_valid():
+            logger.debug("Common name config is not valid, skipping")
+            return
+        config_common_name = self._get_config_common_name()
+        vault.enable_secrets_engine(SecretsBackend.PKI, PKI_MOUNT)
+        if not self._common_name_matches_current_csr(config_common_name):
+            csr = vault.generate_pki_intermediate_ca_csr(
+                mount=PKI_MOUNT, common_name=config_common_name
+            )
+            self._revoke_issued_certificates_and_remove_csr()
+            self.tls_certificates_pki.request_certificate_creation(
+                certificate_signing_request=csr.encode(),
+                is_ca=True,
+            )
+
+    def _revoke_issued_certificates_and_remove_csr(self) -> None:
+        """Revoke all certificates issued by the PKI secrets engine and remove intermediate CA CSR."""
+        csrs = self.tls_certificates_pki.get_requirer_csrs()
+        # There should only be one of them, but for sanity we iterate the list.
+        for csr in csrs:
+            self.tls_certificates_pki.request_certificate_revocation(csr.csr.encode())
+        logger.info("Revoking all certificates issued by the PKI backend.")
+        self.vault_pki.revoke_all_certificates()
+
+    def _is_intermediate_ca_common_name_match(self, vault: Vault, common_name: str) -> bool:
+        """Check if the intermediate CA is set with the valid common name."""
+        intermediate_ca = vault.get_intermediate_ca(mount=PKI_MOUNT)
+        if not intermediate_ca:
+            return False
+        intermediate_ca_common_name = get_common_name_from_certificate(intermediate_ca)
+        return intermediate_ca_common_name == common_name
+
+    def _get_pki_intermediate_ca_csr(self) -> str | None:
+        """Get the current CSR from the relation data."""
+        csrs = self.tls_certificates_pki.get_requirer_csrs()
+        if not csrs:
+            return None
+        assert len(csrs) == 1, f"Only one CSR should be available, found {len(csrs)}: {csrs}"
```

Comment:
What's the intent here? Shouldn't this be logged?

Reply:
This has been refactored.

Reply:
[Here](https://github.com/canonical/vault-k8s-operator/pull/471).
Should I update this branch?

Reply:
No need. Thanks